rfd/0143-external-k8s-joining.md

```diff
@@ -0,0 +1,319 @@
+      projected:
+        sources:
+          - serviceAccountToken:
+              path: my-service-account
+              expirationSeconds: 600
+              audience: example.teleport.sh
+```
+
+### Implementation
+
+The `static_jwks` variant will leverage some of the same parts of the flow
+as the `in_cluster` variant.
+
+1. As with `in_cluster`, the client loads the service account JWT from the
+  filesystem that was mounted by Kubelet and calls `RegisterUsingToken` with 
+  this JWT and the name of the token it wishes to join using.
+2. The Auth Server finds the ProvisionToken matching the name of token. If this
+  is a ProvisionToken with `type: static_jwks`, the flow now diverges from the
+  existing `in_cluster` implementation.
+5. The Auth Server validates:
+  a. The JWT is signed by the keys present in the `jwks` field of the 
+    ProvisionToken.
+  b. The JWT is not expired, and at the time of issue, the JWTs TTL was short.
+    This ensures that users are using the recommended configured TTL value.
+  c. The JWTs audience claim matches the cluster name.
+  d. The JWT includes the `kubernetes.io` claim which binds it to a specified
+    namespace and pod.
+  e. The JWT subject claim is for a service account, and this service account 
+    matches one of those configured in the allow rules of the token resource.
+6. If the JWT passes validation, certificates are signed and returned to the 
+  client.
+
+This flow requires no modifications to the existing RPC protobufs.
+
+## Security Considerations
+
+It's important to note that this join method will be less secure than the 
+`in_cluster` variant. This should be explicitly mentioned in documentation
+for this join method, and encourage the use of `in_cluster` where possible.
+
+This is because the `in_cluster` join method makes use of the TokenReview 
+endpoint which performs additional checks to the validity of the token. This 
+includes checking that the pod and service account listed within the JWT claims 
+exist and are running. This increases the complexity of falsifying a 
+token and also ensures that a token cannot be used beyond the lifetime of the 
+pod it is bound to. We can mimic this by using a very short time to live for 
+the JWT (30 seconds).
+
+However, this join method is still more secure than the long-lived secret
+based `token` join method.
+
+### Token reuse
+
+Token reuse refers to the potential for a malicious actor to intercept
+communications between the client and the server, and to use this interception
+to capture the JWT and then use this elsewhere to obtain credentials.
+
+To mitigate this, a short time-to-live will be used on JWTs issued for the join 
+process. This limits the malicious actors options and forces them to use the JWT
+immediately.
+
+In addition, the audience field is used to ensure that the Teleport Cluster is
+the intended recipient of the JWT and that this has not been stolen from another
+relying party.
```

Comment:
In the end, we've had to switch away from including a nonce in the audience claim. This is because the audience claim can only be configured in the Kubernetes podspec - and therefore must be known by the operator deploying Teleport ahead of time.